pkg/tests/validator_test.go

```diff
@@ -0,0 +1,381 @@
+			0: (uint64(validationBlocksPerSlot) * (epochDurationSlots - 1)) >> uint64(ts.API.ProtocolParameters().SlotsPerEpochExponent()),
+			1: (uint64(validationBlocksPerSlot) * (epochDurationSlots)) >> uint64(ts.API.ProtocolParameters().SlotsPerEpochExponent()),
+			2: (uint64(validationBlocksPerSlot) * (epochDurationSlots)) >> uint64(ts.API.ProtocolParameters().SlotsPerEpochExponent()),
+		},
+	}
+
+	validatorTest(t, test)
+}
+
+func Test_Validator_OverIssuance(t *testing.T) {
+	ts := setupValidatorTestsuite(t)
+	defer ts.Shutdown()
+
+	test := ValidatorTest{
+		ts: ts,
+		// Issue one more block than supposed to.
+		validationBlocksPerSlot: ts.API.ProtocolParameters().ValidationBlocksPerSlot() + 1,
+		epochPerformanceFactors: EpochPerformanceMap{
+			// We expect 0 rewards for overissuance.
+			// We model that in this test by setting performance factor to 0.
+			0: 0,
+			1: 0,
+			2: 0,
+		},
+	}
+
+	validatorTest(t, test)
+}
+
+func Test_Validator_UnderIssuance(t *testing.T) {
+	ts := setupValidatorTestsuite(t)
+	defer ts.Shutdown()
+
+	// Issue less than supposed to.
+	validationBlocksPerSlot := ts.API.ProtocolParameters().ValidationBlocksPerSlot() - 2
+	epochDurationSlots := uint64(ts.API.TimeProvider().EpochDurationSlots())
+
+	test := ValidatorTest{
+		ts:                      ts,
+		validationBlocksPerSlot: validationBlocksPerSlot,
+		epochPerformanceFactors: EpochPerformanceMap{
+			// A validator cannot issue blocks in the genesis slot, so we deduct one slot worth of blocks.
+			0: (uint64(validationBlocksPerSlot) * (epochDurationSlots - 1)) >> uint64(ts.API.ProtocolParameters().SlotsPerEpochExponent()),
+			1: (uint64(validationBlocksPerSlot) * (epochDurationSlots)) >> uint64(ts.API.ProtocolParameters().SlotsPerEpochExponent()),
+			2: (uint64(validationBlocksPerSlot) * (epochDurationSlots)) >> uint64(ts.API.ProtocolParameters().SlotsPerEpochExponent()),
+		},
+	}
+
+	validatorTest(t, test)
+}
+
+func Test_Validator_FixedCostExceedsRewards(t *testing.T) {
+	ts := setupValidatorTestsuite(t, testsuite.WithWalletFixedCost(iotago.MaxMana))
+	defer ts.Shutdown()
+
+	validationBlocksPerSlot := ts.API.ProtocolParameters().ValidationBlocksPerSlot()
+
+	test := ValidatorTest{
+		ts:                      ts,
+		validationBlocksPerSlot: validationBlocksPerSlot,
+		epochPerformanceFactors: EpochPerformanceMap{
+			0: 0,
+			1: 0,
+			2: 0,
+		},
+	}
+
+	validatorTest(t, test)
+}
+
+func validatorTest(t *testing.T, test ValidatorTest) {
+	ts := test.ts
+
+	tip := ts.DefaultWallet().Node.Protocol.Engines.Main.Get().TipSelection.SelectTips(1)[iotago.StrongParentType][0]
+	startEpoch := iotago.EpochIndex(0)
+	endEpoch := iotago.EpochIndex(len(test.epochPerformanceFactors) - 1)
+	fmt.Printf("ValidatorTest: startEpoch=%d, endEpoch=%d\n", startEpoch, endEpoch)
+
+	subslotDuration := time.Duration(ts.API.ProtocolParameters().SlotDurationInSeconds()/test.validationBlocksPerSlot) * time.Second
```

Comment:
This is technically not the subSlotDuration because it is based on the test.validationBlocksPerSlot which is the value used by the validator rather than the protocol param. So this variable should be called issuancePeriod or something along those lines.